README
WebService-HMRC-HelloWorld

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc WebService::HMRC::HelloWorld

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=WebService-HMRC-HelloWorld

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/WebService-HMRC-HelloWorld

    CPAN Ratings
        http://cpanratings.perl.org/d/WebService-HMRC-HelloWorld

    Search CPAN
        http://search.cpan.org/dist/WebService-HMRC-HelloWorld/


LICENSE AND COPYRIGHT

Copyright (C) 2018 Nick Prater

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for WebService::HMRC

0.01    2018-08-06
        First release.



Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'WebService::HMRC',
    AUTHOR           => q{Nick Prater <nick@npbroadcast.com>},
    VERSION_FROM     => 'lib/WebService/HMRC.pm',
    ABSTRACT_FROM    => 'lib/WebService/HMRC.pm',
    LICENSE          => 'artistic_2',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    TEST_REQUIRES => {
        'Test::More' => 0,
        'Test::Exception' => 0,
    },
    PREREQ_PM => {
        'JSON::MaybeXS' => 0,
        'List::Util' => 1.33,
        'LWP::UserAgent' => 0,
        'Moose' => 0,
        'namespace::autoclean' => 0,
        'Try::Tiny' => 0,
        'URI' => 0,
        'URI::Escape' => 0,
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url => 'https://github.com/nick-prater/WebService-HMRC.git',
                web => 'https://github.com/nick-prater/WebService-HMRC',
            },
        },
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'WebService-HMRC-*' },
);


lib/WebService/HMRC.pm
package WebService::HMRC;

use 5.006;
use strict;
use warnings;

=head1 NAME

WebService::HMRC - Interact with the UK HMRC Making Tax Digital API

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

    use WebService::HMRC::Request;

    my $r = WebService::HMRC::Request->new();
    $r->auth->access_token('MY_ACCESS_TOKEN');

    my $response = $r->get_endpoint({
        endpoint => '/hello/user',
        auth_type => 'user',
    });
    print $response->data->{message} if $response->is_success;

=head1 DESCRIPTION

This Perl module is a base for a suite of modules used for
interacting with the HMRC Making Tax Digital (MTD) and Making Tax Digital for
Business (MTDfB) APIs.

HMRC is the UK government tax authority. Their APIs
provide a means to submit and query information relating to personal and
business tax and customs affairs.

These base modules will normally not be used directly by an application.
Instead, applications will generally use a higher-level module which inherits
from these classes.

For more information, see:
L<https://developer.service.hmrc.gov.uk/api-documentation/docs/api>

=head2 CORE MODULES

This module is distributed with the following low-level modules comprising
the essential core classes for interacting with the API:

=over

=item L<WebService::HMRC::Request>

=item L<WebService::HMRC::Response>

=item L<WebService::HMRC::Authenticate>

=back

=head2 API MODULES

Rather than using the low-level modules directly, applications will generally
use a module specific to the API they wish to access (which will inherit from
the core modules). At the time of writing, the following API modules are
available:

=over

=item L<WebService::HMRC::HelloWorld>

=item L<WebService::HMRC::CreateTestUser>

=item L<WebService::HMRC::VAT>

=back

=head1 INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

=head1 AUTHORISATION

Except for a small number of open endpoints, access to the HMRC APIs requires
appliction or user credentials. These must be obtained from HMRC. Application
credentials (and documentation) may be obtained from their
L<Developer Hub|https://developer.service.hmrc.gov.uk/api-documentation>

=head1 TESTING

The basic tests, as run as part of the installation instructions shown above,
do not require an internet connection or authorisation with HMRC.

Developer pre-release tests may be run with the following command:

    RELEASE_TESTING=1 make test

With a working internet connection and appropriate HMRC credentials, specified
as environment variables, interaction with the real HMRC sandbox api, using
authorisation, may be tested:

=head2 Test generation of a valid authorisation url

To run these tests:

=over

=item * An application must be registered with HMRC.

=item * The application must be enabled for the 'Hello World' test api with HMRC.

=item * C<urn:ietf:wg:oauth:2.0:oob> must be set as a valid redirect_uri for the application.

=item * HMRC_CLIENT_ID environment variable must be set (issued when application was registered).

=item * The HMRC sandbox test api endpoints must be functioning.

=back

    HMRC_CLIENT_ID=[MY-CLIENT-ID] prove -l t/04-hmrc-authenticate.t

This test generates a url, shown in the test output, which can be pasted into
a browser to authorise the application and obtain an Authorisation Code for
the following test stage.

=head2 Test token generation and renewal

In addition to the requirements of the test above, the following environment
variables must be set:

=over

=item * HMRC_CLIENT_SECRET (issued when application was registered).

=item * HMRC_AUTH_CODE (obtained after following the authorisation url generated above).

=back

    HMRC_CLIENT_ID=[MY-CLIENT-ID] \
    HMRC_CLIENT_SECRET=[MY-CLIENT_SECRET] \
    HMRC_AUTH_CODE=[MY-AUTH_CODE] \
    prove -l t/04-hmrc-authenticate.t

When successful, this test will output the response from refreshing access
tokens, including new access and refresh tokens.

=head2 Test API Interaction

These tests query the HMRC Hello World API, testing each of the authentication
types and checking for an expected response. Prerequisites must be in place as
for the other tests above and the following environment variables set:

=over

=item * HMRC_SERVER_TOKEN (issued when application was registered)

=item * HMRC_AUTH_CODE (refreshed code output by previous test above)

=back

    HMRC_SERVER_TOKEN=[MY-SERVER-TOKEN] \
    HMRC_AUTH_CODE=[MY-AUTH_CODE] \
    prove -l t/05-hmrc-api.t

=head1 AUTHOR

Nick Prater, <nick@npbroadcast.com>

=head1 BUGS

Please report any bugs or feature requests to C<bug-webservice-hmrc@rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-HMRC>.
I will be notified, and then you'll automatically be notified of progress on your bug as
I make changes.

=head1 SUPPORT AND DOCUMENTATION

After installation, you can find documentation for this module with the
perldoc command.

    perldoc WebService::HMRC

The C<README.pod> file supplied with this distribution is generated from the
L<WebService::HMRC> module's pod by running the following command from the 
distribution root:

    perldoc -u lib/WebService/HMRC.pm > README.pod

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=WebService-HMRC>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/WebService-HMRC>

=item * Search CPAN

L<https://search.cpan.org/dist/WebService-HMRC/>

=item * GitHub

L<https://github.com/nick-prater/WebService-HMRC>

=back

=head1 ACKNOWLEDGEMENTS

This module was originally developed for use as part of the
L<LedgerSMB|https://ledgersmb.org/> open source accounting software.

=head1 LICENSE AND COPYRIGHT

Copyright 2018 Nick Prater, NP Broadcast Limited.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of WebService::HMRC


lib/WebService/HMRC/Response.pm
package WebService::HMRC::Response;

use 5.006;
use Carp;
use JSON::MaybeXS qw(decode_json);
use Moose;
use namespace::autoclean;
use Try::Tiny;


=head1 NAME

WebService::HMRC::Response - Response object for the UK HMRC MTD API

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 DESCRIPTION

This is part of the L<WebService::HMRC> suite of Perl modules for
interacting with the UK's HMRC Making Tax Digital APIs.

This class represents the response from an api call. It is inherited by
other classes which implement bindings to HMRC APIs rather than being
used directly.

=head1 SYNOPSIS

    use WebService::HMRC::HelloWorld;  # inherits this module
    my $hmrc = WebService::HMRC::HelloWorld->new();

    # Hello World endpoint requires no authorisation
    # It returns a WebService::HMRC::Response object reference
    my $hmrc_response = $hmrc->hello_world()

    print "success\n" if $hmrc_response->is_success();

    my $status_code  = $hmrc_response->http->code;
    my $message      = $hmrc_response->data->{message};
    my $content_type = $hmrc_response->header('content-type');
    
=head1 PROPERTIES

=head2 data

Reference to a perl hash representing an api call's json response.

This property is updated on object creation when the http property is set.

If an api call returns no valid json response, this property is
populated with a dummy response, in the same format as the standard
HRMC error response, with code 'INVALID_RESPONSE'. This facilitates
consistent error parsing by the calling application.

=cut

has data => (
    is => 'rw',
    isa => 'Maybe[HashRef]',
    default => undef,
);

=head2 http

The raw HTTP::Response object resulting from an api call. See documentation
for L<HTTP::Response>.

=cut

has http => (
    is => 'ro',
    isa => 'HTTP::Response',
    required => 1,
    trigger => \&_parse_http_response,
    handles => ['is_success', 'header'],
);


=head1 METHODS

=head2 is_success()

Shortcut to the HTTP::Response is_success method.

=head2 header($header_name)

Returns the value of the specified response header, or undef if it does
not exist. Shortcut to the HTTP::Response header method.

=cut


# PRIVATE METHODS

# _parse_http_response($http_response)
#
# Update this class's `data` property to reflect the json content of
# the supplied HTTP::Response object. Returns a reference to the Perl data
# structure representing the json response.

sub _parse_http_response {
    my $self = shift;
    my $http = shift;

    # Under normal conditions, the api always returns json content
    # even for error responses. But unusual error conditions could
    # return content that is invalid json, which will cause the parser
    # to croak. We therefore wrap it in a `try` block.
    my $data = try {
        return decode_json($http->decoded_content);
    }
    catch {
        carp "unable to parse api response as json: $_";

        # Create dummy error message with same format as HMRC
        # error response to facilitate consistent error parsing by
        # calling application.
        return {
            code    => 'INVALID_RESPONSE',
            message => 'No valid JSON data received from api call. '.
                       $http->status_line,
        };
    };

    $self->data($data);
    return $data;
}


=head1 AUTHOR

Nick Prater <nick@npbroadcast.com>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-webservice-hmrc-helloworld@rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-HMRC>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WebService::HMRC::Response


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WebService-HMRC>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WebService-HMRC>

=item * Search CPAN

L<http://search.cpan.org/dist/WebService-HMRC/>

=item * GitHub

L<https://github.com/nick-prater/WebService-HMRC>

=back

=head1 ACKNOWLEDGEMENTS

This module was originally developed for use as part of the
L<LedgerSMB|https://ledgersmb.org/> open source accounting software.

=head1 LICENCE AND COPYRIGHT

Copyright 2018 Nick Prater, NP Broadcast Limited.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

__PACKAGE__->meta->make_immutable;
1;


lib/WebService/HMRC/Authenticate.pm
package WebService::HMRC::Authenticate;

use 5.006;
use Carp;
use List::Util 1.33;
use Moose;
use namespace::autoclean;
use Try::Tiny;
use WebService::HMRC::Response;

extends 'WebService::HMRC::Request';

=head1 NAME

WebService::HMRC::Authenticate - Response object for the UK HMRC HelloWorld API

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

    use WebService:HMRC::Authenticate;

    my $auth = WebService::HMRC::Authenticate->new({
        base_url => 'https://test-api.service.hmrc.gov.uk/',
        client_id => $client_id,
        client_secret => $client_secret,
    });

    # Direct user to this url to authorise our application
    # They will be asked for Government Gateway credentials and
    # to approve access to the specified scope for our application.
    my $url = $auth->authorisation_url({
        scopes => ['read:vat', 'write:vat'],
        redirect_uri => 'urn:ietf:wg:oauth:2.0:oob',
        state => 'session-cookie-hash-or-similar-opaque-value',
    });

    # Once user has authorised our application, an authorisation code
    # is generated. This is either copy/pasted back into our application
    # by the user, or supplied via a callback uri. The
    # authorisation code is valid for 10 minutes.

    # Exchange access code for an access token.
    my $result = $auth->get_access_token({
        access_code => $access_code,
        redirect_uri => 'urn:ietf:wg:oauth:2.0:oob',
    });
    $result->is_success or warn "ERROR: ", $result->data->{message};

    # Has token expired?
    my $expired = ($auth->expires_epoch <= time);

    # The token can be refreshed as long as authority hasn't been revoked
    # for the application and was granted less than 18 months ago.
    # Typically an access token is valid for 4 hours.
    $auth->refresh_tokens;

    # The tokens can be retained and used by another instance
    my $new_auth = WebService::HMRC::Authenticate->new({
        client_id => $client_id,
        client_secret => $client_secret,
        access_token => $access_token,
        refresh_token => $refresh_token,
    })

    # Refreshing will then populate scope and expires_epoch properties
    $new_auth->refresh;

=head1 DESCRIPTION

This module is part of the L<WebService::HMRC> suite of Perl modules for
interacting with the UK's HMRC Making Tax Digital APIs.

It is generally used by other classes which implement bindings to HMRC APIs
rather than being used directly.

This module handles authentication with HMRC using OAuth2. For more detail
see:
L<https://developer.service.hmrc.gov.uk/api-documentation/docs/authorisation/user-restricted-endpoints>

To use this module, an application must first be registered with HMRC, who
will issue client and application credentials.

=head1 INSTALLATION AND TESTING

See documentation for L<WebService::HMRC>.
    
=head1 PROPERTIES

=head2 server_token

Secret server token issued by HMRC for the application using this module. Must be defined
to call application-restricted endpoints. 

=cut

has server_token => (
    is => 'rw',
    isa => 'Str',
    predicate => 'has_server_token',
);

=head2 client_id

The Client ID issued by HMRC for the application using this class.

=cut

has client_id => (
    is => 'rw',
    isa => 'Str',
    predicate => 'has_client_id',
);

=head2 client_secret

The Client Secret issued by HMRC for the application using this class.

=cut

has client_secret => (
    is => 'rw',
    isa => 'Str',
    predicate => 'has_client_secret',
);

=head2 access_token

The access token issued by HMRC. Updated automatically when a new token
is requested using access_token() or refresh_token() methods, or can be
set explicitly to use an existing token.

Access tokens are typically valid for four hours after issue or until
a refreshed token is requested.

=cut

has access_token => (
    is => 'rw',
    isa => 'Str',
    predicate => 'has_access_token',
    clearer => 'clear_access_token',
);

=head2 refresh_token

The refresh token issued by HMRC. Updated automatically when a new token
is requested using access_token() or refresh_token() methods, or can be
set explicitly to use an existing refresh_token.

Refresh tokens are typically valid for 18 months after the application was
last authorised by the user, unless revoked, or until a new refresh token
is requested.

=cut

has refresh_token => (
    is => 'rw',
    isa => 'Str',
    predicate => 'has_refresh_token',
    clearer => 'clear_refresh_token',
);

=head2 expires_epoch

Expiry time of the current access_token and refresh_token specified in seconds
since the perl epoch.

=cut

has expires_epoch => (
    is => 'rw',
    isa => 'Int',
    clearer => 'clear_expires_epoch',
);

=head2 scopes

Arrayref containing a list of scopes for which the current authorisation
tokens are valid. Scopes are specific and documented for each of HMRC's APIs.
For example 'read:vat' or 'write:employment'.

Updated according the the HMRC api response when a new token is requested using
access_token() or refresh_token() methods.

=cut

has scopes => (
    is => 'rw',
    isa => 'ArrayRef',
    predicate => 'has_scopes',
    clearer => 'clear_scopes',
);

=head1 METHODS

=head2 authorisation_url({ scopes => \@scopes, redirect_uri => $uri, [state => $state] })

Returns a URI object representing the url to which the a user should be
directed to authorise this application for the specified scopes. In string
context, this return value evaluates to a fully-qualified url.

=head3 Parameters:

=over

=item scopes

Reference to an array containing the scopes to which access is sought. The scope
names and their application are defined in the documentation for each api.

For example: ['read:employment', 'read:vat', 'write:vat']

=item redirect_uri

Where the user will be redirected once access is granted or denied. This must
have been associated with the application using the HMRC developer web site.
See:
L<https://developer.service.hmrc.gov.uk/api-documentation/docs/reference-guide#redirect-uris>

=item state

Optional parameter. The value specified here is returned as a parameter to the
redirect_uri and should be checked to ensure that the redirect is a genuine
response to our authorisation request.

=back

=cut

sub authorisation_url {

    my ($self, $args) = @_;

    defined $args->{scopes} or croak 'scopes not defined';
    defined $args->{redirect_uri} or croak 'redirect_uri not defined';
    $self->has_client_id or croak 'client_id property not defined for object';

    my $uri = $self->endpoint_url('/oauth/authorize');
    my $scope_string = join(' ', @{$args->{scopes}});
    my $query_params = {
        response_type => 'code',
        client_id => $self->client_id,
        scope => $scope_string,
        redirect_uri => $args->{redirect_uri}
    };

    # state is an optional parameter
    if(defined $args->{state}) {
        $query_params->{state} = $args->{state};
    }

    $uri->query_form($query_params);

    return $uri;
}


=head2 get_access_token({ access_code => $access_code, redirect_uri => $redirect_uri })

Exchanges the supplied access_code for an access_token.

Both client_id and client_secret object properties must be set before calling
this method.

The redirect_uri parameter must match that used when the access_code was 
requested.

Returns a WebService::HMRC::Response object.

The object's scopes, access_token, refresh_token and expires_epoch properties are
updated when this method is called.

See L<https://developer.service.hmrc.gov.uk/api-documentation/docs/authorisation/user-restricted-endpoints>
for more information about response data and possible error codes.

=cut

sub get_access_token {

    my ($self, $args) = @_;

    defined $args->{authorisation_code} or croak 'authorisation_code not defined';
    defined $args->{redirect_uri} or croak 'redirect_uri not defined';
    $self->has_client_id or croak 'client_id property not defined for object';
    $self->has_client_secret or croak 'client_secret property not defined for object';

    my $uri = $self->endpoint_url('/oauth/token');
    my $params = {
        grant_type => 'authorization_code',
        client_id => $self->client_id,
        client_secret => $self->client_secret,
        code => $args->{authorisation_code},
        redirect_uri => $args->{redirect_uri}
    };
    my $http_response = $self->ua->post($uri, $params);
    my $result = WebService::HMRC::Response->new(http => $http_response);
    $self->extract_tokens($result->data);

    return $result;
}


=head2 refresh_tokens()

Exchanges the current tokens for a new access_token

The properties client_id, client_secret object properties must be set before calling
this method.

The object's scopes, access_token, refresh_token and expires_epoch properties are
updated when this method is called.

Returns a WebService::HMRC::Response object. 

See L<https://developer.service.hmrc.gov.uk/api-documentation/docs/authorisation/user-restricted-endpoints>
for more information about response data and possible error codes.

=cut

sub refresh_tokens {

    my $self = shift;

    $self->has_refresh_token or croak 'refresh_token not defined for object';
    $self->has_client_id or croak 'client_id property not defined for object';
    $self->has_client_secret or croak 'client_secret not defined for object';

    my $uri = $self->endpoint_url('/oauth/token');
    my $params = {
        grant_type => 'refresh_token',
        client_id => $self->client_id,
        client_secret => $self->client_secret,
        refresh_token => $self->refresh_token,
    };
    my $http_response = $self->ua->post($uri, $params);
    my $result = WebService::HMRC::Response->new(http => $http_response);
    $self->extract_tokens($result->data);

    return $result;
}


=head2 extract_tokens($hashref)

Accepts a hashref representing a set of tokens, such as the response from
an hmrc token endpoint.

Updates the access_token, refresh_token and expires_epoch properties of this
class.

Returns true on success.

On error, clears the access_token, refresh_token, scopes and expires_epoch
properties and returns false.

A typical token hashref comprises:

    {
      'scope'         => 'read:vat read:employment',# API specific
      'token_type'    => 'bearer',                  # Always 'bearer'
      'expires_in'    => 14400,                     # seconds until expiration
      'refresh_token' => '806d848e5e78fee92c9a38e6b7a3',
      'access_token'  => '7d46efbcbff7892295894e21f940d118'
    }

The token will be rejected if token_type is not 'bearer'.

=cut

sub extract_tokens {

    my $self = shift;
    my $token = shift;

    return try {
        $token->{token_type} eq 'bearer' or croak 'token type value is not `bearer`';
        $token->{expires_in} =~ m/^\d+$/ or croak 'expires_in value is not numeric';

        $self->expires_epoch(time + $token->{expires_in});
        $self->access_token($token->{access_token});
        $self->refresh_token($token->{refresh_token});
        $self->scopes( $self->split_scope_string($token->{scope}) );
        return 1;
    }
    catch {
        carp "Error parsing token: $_\n";
        $self->clear_expires_epoch;
        $self->clear_access_token;
        $self->clear_refresh_token;
        $self->clear_scopes;
        return 0;
    };
}


=head2 split_scope_string($string)

Given a string containing space-delimited authorisation scopes
(e.g. 'read:vat write:employment write_vat'), split into an array
of individual scope names.

Returns an array reference, or undef if the input is undefined.

=cut

sub split_scope_string {

    my $self = shift;
    my $string = shift;

    defined $string or croak 'scopes text string is unefined';

    my @scopes = split / /, $string;
    $self->scopes(\@scopes);

    return \@scopes;
}


=head2 has_scope($scope)

Returns true if the current authorisation tokens are associated with the
specified scope.

The scope of the current tokens is only known (and this method is therefore
only meaningful) after a call to `get_access_token()` or `refresh_tokens()`,
unless the scopes property has been explicitly initialised.

If the scope of the current tokens has not been initialised, this method will
croak.

=cut

sub has_scope {

    my $self = shift;
    my $scope = shift;

    $self->has_scopes
        or croak 'Token scopes are undetermined. Try calling refresh() method before testing scopes';

    return List::Util::any {$_ eq $scope} @{$self->scopes};
}


=head1 AUTHOR

Nick Prater <nick@npbroadcast.com>

=head1 BUGS

Please report any bugs or feature requests to C<bug-webservice-hmrc at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-HMRC>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WebService::HMRC::Authenticate


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WebService-HMRC>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WebService-HMRC>

=item * Search CPAN

L<http://search.cpan.org/dist/WebService-HMRC/>

=back

=head1 ACKNOWLEDGEMENTS

This module was originally developed for use as part of the
L<LedgerSMB|https://ledgersmb.org/> open source accounting software.

=head1 LICENSE AND COPYRIGHT

Copyright 2018 Nick Prater, NP Broadcast Limited.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

__PACKAGE__->meta->make_immutable;
1;
